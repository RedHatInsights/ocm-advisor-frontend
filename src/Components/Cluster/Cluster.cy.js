import React from 'react';
import { mount } from '@cypress/react';
import { MemoryRouter } from 'react-router-dom';

import { Intl } from '../../Utilities/intlHelper';
import { Cluster } from './Cluster';
import { Provider } from 'react-redux';
import getStore from '../../Store';
import singleClusterPageReport from '../../../cypress/fixtures/api/insights-results-aggregator/v2/cluster/dcb95bbf-8673-4f3a-a63c-12d4a530aa6f/reports-disabled-false.json';
import {
  checkNoMatchingRecs,
  checkRowCounts,
} from '../../../cypress/utils/table';

// selectors
const CLUSTER_HEADER = '#cluster-header';
const BREADCRUMBS = 'nav[class=pf-c-breadcrumb]';
const RULES_TABLE = '#cluster-recs-list-table';
const FILTER_CHIPS = 'li[class=pf-c-chip-group__list-item]';
let props;

describe('cluster page', () => {
  beforeEach(() => {
    // the flag tells not to fetch external federated modules
    window.CYPRESS_RUN = true;

    props = {
      cluster: {
        isError: false,
        isUninitialized: false,
        isLoading: false,
        isFetching: false,
        isSuccess: true,
        data: singleClusterPageReport,
      },
      displayName: {
        data: singleClusterPageReport.report.meta.cluster_name,
      },
      clusterId: 'foobar',
    };
  });

  it('cluster page in the successful state', () => {
    mount(
      <MemoryRouter>
        <Intl>
          <Provider store={getStore()}>
            <Cluster {...props} />
          </Provider>
        </Intl>
      </MemoryRouter>
    );
    // renders breadcrumbs
    cy.get(BREADCRUMBS)
      .should('have.length', 1)
      .get('.pf-c-breadcrumb__list > :nth-child(2)')
      .should('have.text', 'Cluster With Issues');
    // renders cluster header
    cy.get(CLUSTER_HEADER).should('have.length', 1);
    // renders table component
    cy.get(RULES_TABLE).should('have.length', 1);
    // test how many rows were rendered
    checkRowCounts(singleClusterPageReport.report.data.length, true);
  });

  it('cluster page in the loading state', () => {
    props = {
      ...props,
      cluster: {
        ...props.cluster,
        isFetching: true,
        isSuccess: false,
        data: undefined,
      },
    };
    mount(
      <MemoryRouter>
        <Intl>
          <Provider store={getStore()}>
            <Cluster {...props} />
          </Provider>
        </Intl>
      </MemoryRouter>
    );
    // renders breadcrumbs
    cy.get(BREADCRUMBS).should('have.length', 1);
    // renders cluster header
    cy.get(CLUSTER_HEADER).should('have.length', 1);
    // renders table component
    cy.get(RULES_TABLE).should('have.length', 1);
    cy.ouiaId('loading-skeleton').should('have.length', 1);
  });

  it('cluster page in the error state', () => {
    props = {
      ...props,
      cluster: {
        ...props.cluster,
        isError: true,
        isSuccess: false,
        isFetching: false,
        data: undefined,
      },
    };
    mount(
      <MemoryRouter>
        <Intl>
          <Provider store={getStore()}>
            <Cluster {...props} />
          </Provider>
        </Intl>
      </MemoryRouter>
    );
    // renders breadcrumbs
    cy.get(BREADCRUMBS).should('have.length', 1);
    // renders cluster header
    cy.get(CLUSTER_HEADER).should('have.length', 1);
    // renders table component
    cy.get(RULES_TABLE).should('have.length', 1);
    cy.get('.pf-c-empty-state').should('have.length', 1);
  });

  it('adds additional filters passed by the query parameters №1', () => {
    mount(
      <MemoryRouter initialEntries={['?total_risk=1&text=foo+bar&category=2']}>
        <Intl>
          <Provider store={getStore()}>
            <Cluster {...props} />
          </Provider>
        </Intl>
      </MemoryRouter>
    );
    cy.get(BREADCRUMBS);
    cy.get(CLUSTER_HEADER);
    cy.get(RULES_TABLE);
    cy.get(FILTER_CHIPS).each(($el) =>
      expect($el.text()).to.be.oneOf(['foo bar', 'Low', 'Performance'])
    );
    checkNoMatchingRecs();
  });

  it('adds additional filters passed by the query parameters №2', () => {
    mount(
      <MemoryRouter initialEntries={['?total_risk=2&text=foo&category=1']}>
        <Intl>
          <Provider store={getStore()}>
            <Cluster {...props} />
          </Provider>
        </Intl>
      </MemoryRouter>
    );
    cy.get(BREADCRUMBS);
    cy.get(CLUSTER_HEADER);
    cy.get(RULES_TABLE);
    cy.get(FILTER_CHIPS).each(($el) =>
      expect($el.text()).to.be.oneOf([
        'foo',
        'Moderate',
        'Service Availability',
      ])
    );
    checkNoMatchingRecs();
  });
});

describe('Cluster page display name test №1', () => {
  before(() => {
    props = {
      cluster: {
        isError: false,
        isUninitialized: false,
        isLoading: false,
        isFetching: false,
        isSuccess: true,
        data: singleClusterPageReport,
      },
      clusterId: 'Cluster Id',
    };
  });

  it('Cluster breadcrumbs name should be Cluster With Issues', () => {
    mount(
      <MemoryRouter>
        <Intl>
          <Provider store={getStore()}>
            <Cluster {...props} />
          </Provider>
        </Intl>
      </MemoryRouter>
    );
    cy.get(BREADCRUMBS)
      .should('have.length', 1)
      .get('.pf-c-breadcrumb__list > :nth-child(2)')
      .should('have.text', 'Cluster With Issues');
  });

  it('Cluster breadcrumbs name should be = Cluster Id', () => {
    mount(
      <MemoryRouter>
        <Intl>
          <Provider store={getStore()}>
            <Cluster
              {...{ ...props, cluster: { ...props.cluster, data: null } }}
            />
          </Provider>
        </Intl>
      </MemoryRouter>
    );
    cy.get(BREADCRUMBS)
      .should('have.length', 1)
      .get('.pf-c-breadcrumb__list > :nth-child(2)')
      .should('have.text', 'Cluster Id');
  });
});
