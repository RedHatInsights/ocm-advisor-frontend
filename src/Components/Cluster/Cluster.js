import './_Cluster.scss';

import React, { useEffect } from 'react';
import { useIntl } from 'react-intl';
import PropTypes from 'prop-types';
import { useDispatch, useSelector } from 'react-redux';
import { Grid, GridItem } from '@patternfly/react-core';

import { SearchIcon } from '@patternfly/react-icons';
import PageHeader from '@redhat-cloud-services/frontend-components/PageHeader';
import Main from '@redhat-cloud-services/frontend-components/Main';
import routerParams from '@redhat-cloud-services/frontend-components-utilities/RouterParams';

import messages from '../../Messages';
import ClusterHeader from '../ClusterHeader/ClusterHeader';
import ClusterRules from '../ClusterRules/ClusterRules';
import Breadcrumbs from '../Breadcrumbs/Breadcrumbs';
import { fetchClusterById } from '../../AppActions';
import MessageState from '../MessageState/MessageState';
import Loading from '../Loading/Loading';

const Cluster = ({ match }) => {
  const intl = useIntl();
  const dispatch = useDispatch();
  const cluster = useSelector(
    (AdvisorStore) => AdvisorStore.clusters[match.params.clusterId]
  );

  useEffect(() => {
    if (match.params.clusterId) {
      const subnav = `${match.params.clusterId} - ${intl.formatMessage(
        messages.clusters
      )}`;
      document.title = intl.formatMessage(messages.documentTitle, { subnav });
    }

    dispatch(fetchClusterById(match.params.clusterId));
  }, [match.params.clusterId]);

  return (
    <React.Fragment>
      <PageHeader className="pf-m-light ins-inventory-detail">
        {cluster && (
          <Breadcrumbs current={match.params.clusterId} match={match} />
        )}
        <ClusterHeader uuid={match.params.clusterId} />
      </PageHeader>
      <Main>
        <React.Fragment>
          {cluster && cluster.fetchStatus === 'rejected' && (
            <MessageState
              title="No recommendations available"
              text="There was an error fetching recommendations for this cluster. Refresh your page to try again."
              icon={SearchIcon}
            />
          )}
          {cluster && cluster.fetchStatus === 'pending' && <Loading />}
          {cluster && cluster.fetchStatus === 'fulfilled' && (
            <Grid hasGutter>
              <GridItem span={12}>
                <ClusterRules cluster={cluster} />
              </GridItem>
            </Grid>
          )}
        </React.Fragment>
      </Main>
    </React.Fragment>
  );
};

Cluster.propTypes = {
  history: PropTypes.object,
  match: PropTypes.any,
};

export default routerParams(Cluster);
