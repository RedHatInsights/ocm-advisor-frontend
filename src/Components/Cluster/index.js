import React, { useEffect } from 'react';
import routerParams from '@redhat-cloud-services/frontend-components-utilities/RouterParams';
import { useIntl } from 'react-intl';

import { useGetClusterByIdQuery } from '../../Services/SmartProxy';
import messages from '../../Messages';
import { Cluster } from './Cluster';
import { useGetClusterDisplayNameByIdQuery } from '../../Services/AccountManagementService';

export default routerParams(({ match }) => {
  const intl = useIntl();
  const cluster = useGetClusterByIdQuery(match.params.clusterId);
  const displayName = useGetClusterDisplayNameByIdQuery(match.params.clusterId);

  useEffect(() => {
    cluster.refetch();
  }, [match.params.clusterId]);

  useEffect(() => {
    if (match.params.clusterId) {
      const subnav = `${match.params.clusterId} - ${intl.formatMessage(
        messages.clusters
      )}`;
      document.title = intl.formatMessage(messages.documentTitle, { subnav });
    }
  }, [match.params.clusterId]);

  return <Cluster cluster={cluster} displayName={displayName} match={match} />;
});
