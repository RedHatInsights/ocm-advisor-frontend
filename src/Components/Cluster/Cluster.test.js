import { shallow } from 'enzyme';
import React from 'react';

import { Cluster } from './Cluster';

jest.mock('../ClusterHeader', () => jest.fn());

describe('<Cluster /> test', () => {
  let props;

  beforeEach(() => {
    props = {
      cluster: {
        isError: false,
        isUninitialized: false,
        isLoading: false,
        isFetching: false,
        isSuccess: true,
        data: {},
      },
      displayName: {
        data: 'display-name-123',
      },
      match: {
        params: {
          clusterId: 'foobar',
        },
        url: 'foobar',
      },
    };
  });
  test('renders correctly', () => {
    const shallowed = shallow(<Cluster {...props} />);
    expect(shallowed).toMatchSnapshot();
  });
  test('renders correctly when cluster loading', () => {
    props.cluster.isFetching = true;
    props.cluster.isSuccess = false;
    const shallowed = shallow(<Cluster {...props} />);
    expect(shallowed).toMatchSnapshot();
  });
  test('renders correctly when cluster unavailable', () => {
    props.cluster.isError = true;
    props.cluster.isSuccess = false;
    const shallowed = shallow(<Cluster {...props} />);
    expect(shallowed).toMatchSnapshot();
  });
});
