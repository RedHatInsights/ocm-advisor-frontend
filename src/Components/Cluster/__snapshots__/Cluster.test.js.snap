// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cluster /> test renders correctly 1`] = `
<Fragment>
  <PageHeader
    className="pf-m-light ins-inventory-detail"
  >
    <Breadcrumbs
      current="display-name-123"
      match={
        Object {
          "params": Object {
            "clusterId": "foobar",
          },
          "url": "foobar",
        }
      }
    />
    <mockConstructor />
  </PageHeader>
  <Connect(Main)>
    <Grid
      hasGutter={true}
    >
      <GridItem
        span={12}
      >
        <ClusterRules
          reports={Array []}
        />
      </GridItem>
    </Grid>
  </Connect(Main)>
</Fragment>
`;

exports[`<Cluster /> test renders correctly when cluster loading 1`] = `
<Fragment>
  <PageHeader
    className="pf-m-light ins-inventory-detail"
  >
    <Breadcrumbs
      current="display-name-123"
      match={
        Object {
          "params": Object {
            "clusterId": "foobar",
          },
          "url": "foobar",
        }
      }
    />
    <mockConstructor />
  </PageHeader>
  <Connect(Main)>
    <Loading />
  </Connect(Main)>
</Fragment>
`;

exports[`<Cluster /> test renders correctly when cluster unavailable 1`] = `
<Fragment>
  <PageHeader
    className="pf-m-light ins-inventory-detail"
  >
    <Breadcrumbs
      current="display-name-123"
      match={
        Object {
          "params": Object {
            "clusterId": "foobar",
          },
          "url": "foobar",
        }
      }
    />
    <mockConstructor />
  </PageHeader>
  <Connect(Main)>
    <MessageState
      icon={[Function]}
      iconStyle={
        Object {
          "color": "#c9190b",
        }
      }
      text="There was an error fetching recommendations for this cluster. Refresh your page to try again."
      title="No recommendations available."
      variant="full"
    />
  </Connect(Main)>
</Fragment>
`;
